Adds radare2, libuv, libzip, xxhash and capstone packages.

roobie opened the pull request.

roobie committed:
Adds radare2, libuv, libzip, xxhash and capstone packages.

sogaiu commented:
I'll start taking a look.

sogaiu commented on community/xxhash.hpkg:
```diff
@@ -0,0 +1,29 @@
+     (os/setenv "PATH"
+                (join-pkg-paths ":" "/bin"
+                                [core/coreutils core/gcc core/make
+                                 core/sed core/which]))
```
I tried leaving out `core/which` and it seemed to build and run ok.

Is it necessary?

roobie replied:
You are correct, it isn't needed.

sogaiu commented on community/xxhash.hpkg:
```diff
@@ -0,0 +1,29 @@
+     (unpack-src xxhash-src)
+     (core/link-/bin/sh)
+     (sh/$ make
+           PREFIX= ^ (dyn :pkg-out)
```
It seemed to work without this PREFIX.

The latter one seemed necessary though.

Did you need the first one too?

roobie replied:
Nope, wasn't needed.

roobie committed:
resolve incorrect builder for xxhash

sogaiu commented on community/radare2.hpkg:
```diff
@@ -0,0 +1,124 @@
+    (os/setenv "CFLAGS"
+               (string *default-cflags*
+                       " "
+                       "-I" (capstone :path) "/include"
```
I started using the following sort of thing: https://gist.github.com/sogaiu/03be65dd99c0efea5d1fc54bcec4b6e9#file-st-hpkg-L120_L143

May be that or something similar is applicable in this file in more than one place?

sogaiu commented on community/radare2.hpkg:
```diff
@@ -0,0 +1,124 @@
+    (os/setenv "CFLAGS"
+               (string *default-cflags*
+                       " "
+                       "-I" (capstone :path) "/include"
```
I started using the following sort of thing: https://gist.github.com/sogaiu/03be65dd99c0efea5d1fc54bcec4b6e9#file-st-hpkg-L120_L143

May be that or something similar is applicable in this file in more than one place?

roobie replied:
Yeah, I that's much more readable. I'll look into transforming it.

roobie replied:
I've updated the code (also re-ordered the packages so that they are in the same order on all places)

roobie commented on community/radare2.hpkg:
```diff
@@ -0,0 +1,124 @@
+
+    (os/setenv "LD_LIBRARY_PATH"
+               (string
+                (join-pkg-paths ":" "/lib"
```
This must be wrong. Could be that LD_LIBRARY_PATH isn't needed. I'm testing now

roobie replied:
Worked without it, so removed.

roobie committed:
Resolve issues, clean up

Implement suggestion using string/join over map for long lists of
package references.

Remove LD_LIBRARY_PATH os env var, because it did no difference.

sogaiu commented on community/radare2.hpkg:
```diff
@@ -0,0 +1,105 @@
+                                          libuv
+                                          libzip]) " ")
+                       " "
+                       (sh/$<_ pkg-config --cflags capstone "liblz4" zlib libuv libxxhash openssl)))
```
Is there a reason for liblz4 being in double quotes?

sogaiu replied:
Line 88 has a similar instance...we really like liblz4 for being so fast :)

roobie replied:
I guess this warrants a comment, but what I observed was that the `.pc` file for `lz4` is called `liblz4.pc` so for `pkg-config` to find the configuration, the argument in the invocation has to have that name. E.g. `/hpkg/345928850d3e62e09f254d4b68f2dbd119300901-lz4/lib/pkgconfig/liblz4.pc`

roobie replied:
ah, but it need not be in quotes despite that, right?

sogaiu replied:
Yes, I think that is correct.

roobie replied:
quotes removed, build works

sogaiu commented on community/libzip.hpkg:
```diff
@@ -0,0 +1,66 @@
+                        " "
+                        "-I" (zlib :path) "/include"
+                        " "
+                        (sh/$<_ pkg-config --cflags zlib)))
```
I think mostly we've been using `sh/$<_` for invocations of `pkg-config`.  I think that version will trim whitespace: https://github.com/andrewchambers/janet-sh/blob/master/sh.janet#L247_L256

I know I haven't done it in every case.  Another item for a linter perhaps...

sogaiu commented on community/libzip.hpkg:
```diff
@@ -0,0 +1,66 @@
+           # in this case, cmake does not seem to honour the -l and -L switches fully, so
+           -DZLIB_INCLUDE_DIR= ^ (string (zlib :path) "/include")
+           -DZLIB_LIBRARY= ^ (string (zlib :path) "/lib")
+           -DCMAKE_SKIP_RPATH=TRUE # ie. skip stripping, I think
```
According to: https://cmake.org/cmake/help/v3.17/variable/CMAKE_SKIP_RPATH.html

> If this is set to TRUE, then the rpath information is not added to compiled executables. The default is to add rpath information if the platform supports it.

I think what that means in practice is that cmake might tamper with the RPATH (not sure if it applies to RUNPATH as well) information, so that having TRUE means it won't mess with what is already there.

Perhaps some experiments are in order.

sogaiu replied:
I tried building without this line and it seems ok.

Perhaps you wouldn't mind confirming?

roobie replied:
I will try it out

roobie replied:
New commit with out this line, and also without the explicit `-I` and `-L`. ELF looks good and binaries run fine without it.

sogaiu commented on community/libuv.hpkg:
```diff
@@ -0,0 +1,44 @@
+    (os/setenv "LDFLAGS" *default-ldflags*)
+
+    # XXX: unsure if actually needed
+    (os/setenv "PREFIX" (dyn :pkg-out))
```
I tried building with this line commented and it appears to have succeeded without error.

roobie replied:
Yes, you are correct. Updated.

sogaiu commented on community/libzip.hpkg:
```diff
@@ -0,0 +1,66 @@
+                (string *default-ldflags*
+                        " "
+                        "-L" (zlib :path) "/lib"
+                        " "
```
I inserted `"-Wl,--enable-new-dtags"` with appropriate spaces padding around here to get the resulting executables and shared objects to use RUNPATH.  One example of this type of thing here: https://github.com/andrewchambers/hpkgs/blob/master/community/proot.hpkg#L40

AFAICT, it's not always necessary though.  For example, tmux's .hpkg doesn't have it: https://github.com/andrewchambers/hpkgs/blob/master/community/tmux.hpkg -- and when I check tmux's executable using `readelf -d`, I see it uses `RUNPATH`, not `RPATH`.

roobie replied:
I updated with your suggestion, and it now uses RUNPATH instead of RPATH

sogaiu commented on community/libzip.hpkg:
```diff
@@ -0,0 +1,66 @@
+                                 core/sed
+                                 cmake]))
+
+     (os/setenv "CFLAGS"
```
I don't know why it works, but just using `*default-cflags*` seems to produce appropriate results.

Removing line 41 `"-L" (zlib :path) "/lib"` seems to succeed too, but not `(sh/$<_ pkg-config --libs zlib)` from line 49.

May be it's good to leave the lines in but possibly make comments that they are not necessary?

roobie replied:
I removed the `-L` and `-I` switches, because they (should) come from the invocation of `pkg-config`

roobie replied:
E.g.
```
PKG_CONFIG_PATH=/hpkg/59038568ddfdc0abe4b0143b20dfe952068df01a-zlib/lib/pkgconfig pkg-config --cflags zlib
=> -I/hpkg/59038568ddfdc0abe4b0143b20dfe952068df01a-zlib/include
```

sogaiu replied:
I've pulled from the branch and successfully built + run radare2 so it seems to be working here :)

sogaiu commented on community/radare2.hpkg:
```diff
@@ -0,0 +1,105 @@
+                       " "
+                       "-lmagic" # doesn't have a .pc
+                       " "
+                       (sh/$<_ pkg-config --libs capstone zlib "liblz4" libxxhash openssl libuv libzip)))
```
To get openssl stuff used, I think one needs to do things a bit differently: https://github.com/andrewchambers/hpkgs/blob/master/community/python3.hpkg#L45

Specifically, use `libssl` and `libcrypto`, not `openssl`.  This only applies to `pkg-config` invocations in this file IIUC (so also line 61).

It may be obvious, but I think for pkg-config we're asking about specific libraries (of which there can be more than one per "package" -- in this case, openssl provides more than one library and none of them is called `openssl`).

I've found this can get confusing because often enough the package name and the library name match, but this may be worth noting as a potential gotcha.

sogaiu replied:
I don't know if there was error output from using `openssl` (I didn't notice, but then I didn't look too closely either).

sogaiu replied:
I found this in the pkg-config man page:
> --print-errors
> If one or more of the modules on the command line, or 
> their dependencies, are not found, or if an error occurs in 
> parsing a .pc file, then this option will cause errors  
> explaining the problem to be printed.

May be this is worth considering for all .hpkg files that use `pkg-config`.

roobie replied:
I see what you mean, but I think it did work, because there's explicitly `-I` and `-L` for openssl, whereas the `-l` have been added by the `pkg-config openssl --[libs|cflags]`. Example, on my machine:
`pkg-config openssl --cflags --print-errors # result: empty`
`pkg-config openssl --libs --print-errors # result: -lssl -lcrypto`
or with the hpkg:
```
PKG_CONFIG_PATH=/hpkg/<hash>-openssl/lib/pkgconfig pkg-config openssl --cflags
-I/hpkg/<hash>-openssl/include 

PKG_CONFIG_PATH=/hpkg/<hash>-openssl/lib/pkgconfig pkg-config openssl --libs  
-L/hpkg/<hash>-openssl/lib -lssl -lcrypto 

PKG_CONFIG_PATH=/hpkg/<hash>-openssl/lib/pkgconfig pkg-config libssl --libs
-L/hpkg/<hash>-openssl/lib -lssl 

PKG_CONFIG_PATH=/hpkg/<hash>-openssl/lib/pkgconfig pkg-config libcrypto --libs
-L/hpkg/<hash>-openssl/lib -lcrypto 
```
Nevertheless, it might be better to do it like in `python3.hpkg` because then we can drop the explicit `-I`s and `-L`s.

sogaiu replied:
Ah, sorry.  I was mistaken.

From the pkg-config man page:

>  The package name specified on the pkg-config command line is defined to be the name of the metadata file, minus the .pc extension.

Here, doing `ls /hpkg/*openssl/lib/pkgconfig/` yields:
```
libcrypto.pc  libssl.pc  openssl.pc
```

It does make sense to use libcrypto and libssl to get the flags we are after.  However, if we were to also use `LDLIBS`, using `openssl` with `pkg-config --libs-only-l`, we'd get just the `-l` info for libcrypto and libssl IIUC.  If we did that though, we'd probably want to do use `pkg-config --libs-only-L --libs-only-other` for `LDFLAGS`.

This kind of detailed splitting seems to matter for some packages and I wasn't aware of it until very recently.  It's not done yet, but here's an example I'm working on that has an example: https://gist.github.com/sogaiu/fbaada1fa249c5dc013b7913e56715df

I think I need to study the details of .pc files more :)

sogaiu replied:
If we don't need to use LDLIBS for this case, let's not do so.  We can modify it if necessary later.

roobie replied:
Then I am leaving it as is, here, IIUC?

sogaiu replied:
Yes, I think for the moment that makes sense, thanks!

(This was the first time I tried github's code review and I think it was kind of confusing -- how did you find it?)

roobie replied:
I agree with you somewhat. I think I should've polished the packages a bit more before I opened the PR, because with this many comments, it gets hard to follow. But also perhaps I should've opened separate PRs for each package. That would've focused the subject matter of the PR and probably make it easier to discuss and reason about. Do you think it would be a good idea to recommend 1 package=1 PR? Or other thoughts?

sogaiu replied:
I think the one package file per PR thing is probably worth trying.

I say "file" because some .hpkg files have more than one package defintion (actually most have at least two because defsrc is actually defining a package too).

sogaiu replied:
Having said that, for this PR, I'm ok with it remaining in its current state.

roobie replied:
Good point. Agreed.

sogaiu commented on community/radare2.hpkg:
```diff
@@ -0,0 +1,105 @@
+          --with-openssl)
+
+    # XXX: seems like it the makefiles assumes that /lib exists
+    (os/mkdir (string (dyn :pkg-out) "/lib"))
```
I tried building without this line and didn't have any difficulties -- that is, I looked at the produced result and noticed that there was a populated `lib` directory.

May be that didn't occur for you?

roobie replied:
You're correct. I think that what I saw was an error caused by something else, which is not there anymore.

sogaiu commented on community/radare2.hpkg:
```diff
@@ -0,0 +1,105 @@
+                       (sh/$<_ pkg-config --cflags capstone "liblz4" zlib libuv libxxhash openssl)))
+    (os/setenv "LDFLAGS"
+               (string *default-ldflags*
+                       " "
```
I noticed the build results used `RPATH` so I inserted a space-padded `-Wl,--enable-new-dtags` around here.

The result was that `RUNPATH` was used in the resulting executables and shared objects.

roobie replied:
Fixed in 8bf0901

roobie committed:
Clean up `libzip`, removing unnecessary configuration.

I have validated the ELF and the binaries seem to run fine.

roobie committed:
Add --with-new-dtags to `libzip`

so that the binaries use RUNPATH instead of RPATH

roobie committed:
Remove unnecessary env-var

roobie committed:
Cleans up radare2.hpkg.

Makes better use of output from `pkg-config`, both with
regards to openssl, and more generally that if a package
has a working .pc, we don't need to explicitly pass the
-I and -L flags to the compiler/linker.

Makes use of "-Wl,--enable-new-dtags" as can be seen in the
snippet below

Exmample snippet from `readelf -d` on bin/r2
0x000000000000001d (RUNPATH)            Library runpath: [/hpkg/d5aa390d0da343c9752dd580ea2199209d687619-radare2/lib:/hpkg/5ac0042cc743e9dfe7ea02f21c59809d191452e9-capstone/lib:/hpkg/280b72d2a9ef0c97f0059277c37873b21c882a39-file/lib:/hpkg/617251c7be2c6870231c5b9cb6958fd8d02ab1ca-zlib/lib:/hpkg/345928850d3e62e09f254d4b68f2dbd119300901-lz4/lib:/hpkg/f696cb6d78a3920e083c859e25a5e515f42e707d-xxhash/lib:/hpkg/212852e8cc98046af28410727bc94b097df9578d-openssl/lib:/hpkg/e6e052d4a123406cd422a0ba8ffed78ac0b27895-libuv/lib64:/hpkg/71ba6f0c3119bb6ade3bf97ecd7e6814469440f8-libzip/lib64]

roobie committed:
fix placement of end brackets

roobie commented on community/radare2.hpkg:
```diff
@@ -0,0 +1,92 @@
+                       " "
+                       (string "-I" (file :path) "/include")
+                       " "
+                       # capstone does this weird thing that there is a `capstone` directory in the
```
I'm not sure if this long comment here is warranted. Should I remove it?

sogaiu replied:
I think it's not bad to document this type of thing.

sogaiu replied:
I think I encountered a similar situation for ncurses and possibly other things.

I haven't convinced myself why the issue exists yet -- may be there is something I could do to get things working appropriately.

May be we could name such situations (e.g. unaligned include path or something), then the comment wouldn't have to be long.  Though perhaps we'd want some place for a reader to be able to look to learn what the name means.

sogaiu replied:
If we take notes / log about situations we encountered, perhaps these can be organized and referred to.  This ought to help us in future work possibly also for potential contributors.

I was thinking that the topic for a package in the intent-to-package stream might not be a bad place to take some notes.

sogaiu replied:
I'm fine with the long comment for now.  I think it can be shortened or removed later.

roobie replied:
I noted this under the `capstone` package in the `intent-to-package` stream.

andrewchambers commented:
This ok to merge?

sogaiu commented:
It looks pretty good to me :)

andrewchambers commented:
Thanks! I've been a bit busy recently.
